### PR TITLE
feat(datasets) Add VerticalEvenPartitioner

### PR DESCRIPTION
**Pull request by adam-narozniak**
A new class that produces even or approximately even number of partitions (in terms of number of columns).

This class can't be a subclass of `VerticalSizePartitioner` due to the inability of accurate fraction (neither count) specification of sizes. Why?
e.g. 
```
n_cols = 104
columns = [f"{i}" for i in range(n_cols)]
dataset = _create_dummy_dataset(columns, num_rows=50)
# If using the subclassing
partitioner = VerticalEvenPartitioner(num_partitions=100, shuffle=False)
partitioner.dataset = dataset
pd.Series(
    [partitioner.load_partition(i).num_columns for i in range(100)]
).value_counts()
# 1    99
# 5     1
# Name: count, dtype: int64
```

And here if using a new class
```
1    96
2     4
Name: count, dtype: int64
```
```
0     2
1     2
2     2
3     2
4     1
     ..
95    1
96    1
97    1
```

So why it is not possible to inherit?
an operation that creates either counts of fractions needs to happen at the init time (to pass the args to SizePartitinoer) However due to the late assignment of the dataset it's not possible to use count neither to provide a better way of assign fractions that [1/num_partitions] * num_partitions and the assign the remainder columns to the signed object.

### CODE REVIEW

File: datasets/flwr_datasets/partitioner/vertical_even_partitioner.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+from flwr_datasets.partitioner.vertical_partitioner_utils import (
+    _add_active_party_columns,
+    _list_split,
+)
+
+
+class VerticalEvenPartitioner(Partitioner):
+    """Partitioner that splits features (columns) evenly into vertical partitions.
+
+    Enables selection of "active party" column(s) and palcement into
```
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/vertical_even_partitioner.py:33]:
```suggestion
    Enables selection of "active party" column(s) and placement into
```

File: datasets/flwr_datasets/partitioner/vertical_even_partitioner.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    @property
+    def num_partitions(self) -> int:
+        """Number of partitions."""
+        self._determine_partitions_if_needed()
+        assert self._partition_columns is not None
+        return len(self._partition_columns)
+
+    def _validate_parameters_in_init(self) -> None:
+        if self._num_partitions < 1:
+            raise ValueError("column_distribution as int must be >= 1.")
```
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/vertical_even_partitioner.py:178]:
```suggestion
            raise ValueError("`column_distribution` as int must be >= 1.")
```

File: datasets/flwr_datasets/partitioner/vertical_even_partitioner.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+            "create_as_first",
+            "create_as_last",
+            "add_to_all",
+        }
+        if not (
+            isinstance(self._active_party_columns_mode, int)
+            or self._active_party_columns_mode in valid_modes
+        ):
+            raise ValueError(
+                "active_party_column_mode must be an int or one of "
```
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/vertical_even_partitioner.py:201]:
```suggestion
                "`active_party_column_mode` must be an int or one of "
```

File: datasets/flwr_datasets/partitioner/vertical_even_partitioner.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+            raise ValueError("column_distribution as int must be >= 1.")
+
+        # Validate columns lists
+        for parameter_name, parameter_list in [
+            ("drop_columns", self._drop_columns),
+            ("shared_columns", self._shared_columns),
+            ("active_party_column", self._active_party_columns),
+        ]:
+            if not all(isinstance(column, str) for column in parameter_list):
+                raise ValueError(f"All entries in {parameter_name} must be strings.")
```
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/vertical_even_partitioner.py:187]:
```suggestion
                raise ValueError(f"All entries in '{parameter_name}' must be strings.")
```

File: datasets/flwr_datasets/partitioner/vertical_even_partitioner_test.py [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        """Test initialization with non-string elements in shared_columns."""
+        with self.assertRaises(ValueError):
+            VerticalEvenPartitioner(num_partitions=2, shared_columns=["col1", 123])
+
+    def test_init_with_non_string_active_party_column(self) -> None:
+        """Test initialization with non-string elements in active_party_column."""
+        with self.assertRaises(ValueError):
+            VerticalEvenPartitioner(
+                num_partitions=2, active_party_columns=["col1", None]
+            )
```
Comment (jafermarq) [inline datasets/flwr_datasets/partitioner/vertical_even_partitioner_test.py:66]:
```suggestion
        def test_init_with_non_list_drop_columns(self) -> None:
        """Test initialization with non-list elements in drop_columns."""
        with self.assertRaises(ValueError):
            VerticalEvenPartitioner(num_partitions=2, drop_columns="a")

    def test_init_with_non_string_shared_columns(self) -> None:
        """Test initialization with non-string elements in shared_columns."""
        with self.assertRaises(ValueError):
            VerticalEvenPartitioner(num_partitions=2, shared_columns=["col1", 123])

    def test_init_with_non_list_shared_columns(self) -> None:
        """Test initialization with non-list elements in shared_columns."""
        with self.assertRaises(ValueError):
            VerticalEvenPartitioner(num_partitions=2, shared_columns="col1")

    def test_init_with_non_string_active_party_column(self) -> None:
        """Test initialization with non-string elements in active_party_column."""
        with self.assertRaises(ValueError):
            VerticalEvenPartitioner(
                num_partitions=2, active_party_columns=["col1", None]
            )

    def test_init_with_non_list_active_party_column(self) -> None:
        """Test initialization with non-list elements in active_party_column."""
        with self.assertRaises(ValueError):
            VerticalEvenPartitioner(num_partitions=2, active_party_columns="col1")
```

File: datasets/flwr_datasets/partitioner/vertical_even_partitioner.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+            raise ValueError("`column_distribution` as int must be >= 1.")
+
+        # Validate columns lists
+        for parameter_name, parameter_list in [
+            ("drop_columns", self._drop_columns),
+            ("shared_columns", self._shared_columns),
+            ("active_party_column", self._active_party_columns),
+        ]:
+            if not all(isinstance(column, str) for column in parameter_list):
+                raise ValueError(f"All entries in '{parameter_name}' must be strings.")
```
[the remark this answers is not in the document]
Comment (adam-narozniak) [reply datasets/flwr_datasets/partitioner/vertical_even_partitioner.py]: moved to a new init function

**Review by jafermarq — DISMISSED** [review]

**Review by jafermarq — APPROVED** [review]

**Pull request merged**